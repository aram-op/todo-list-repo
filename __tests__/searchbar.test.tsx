import {fireEvent, render, screen, waitFor} from '@testing-library/react';
import {usePathname, useRouter, useSearchParams} from 'next/navigation';
import React from 'react';
import Searchbar from '@/app/ui/searchbar/searchbar';

jest.mock('next/navigation', () => ({
    usePathname: jest.fn(),
    useRouter: jest.fn(),
    useSearchParams: jest.fn(),
}));

jest.mock('@/app/ui/fonts', () => ({
    poppins: {className: 'mock-poppins-font'},
}));

describe('Searchbar component', () => {
    const mockReplace = jest.fn();

    beforeEach(() => {
        jest.clearAllMocks();
        (useRouter as jest.Mock).mockReturnValue({replace: mockReplace});
        (usePathname as jest.Mock).mockReturnValue('/todos');
        (useSearchParams as jest.Mock).mockReturnValue(new URLSearchParams('?query=existing'));
    });

    it('should render the search input and button', () => {
        render(<Searchbar/>);

        const input = screen.getByRole('textbox');
        const button = screen.getByRole('button');

        expect(input).toBeInTheDocument();
        expect(input).toHaveClass('mock-poppins-font');
        expect(button).toBeInTheDocument();
    });

    it('should update the query parameter if input changes', async () => {
        render(<Searchbar/>);

        const input = screen.getByRole('textbox');
        fireEvent.change(input, {target: {value: 'mock filter text'}});

        await waitFor(() => {
            expect(mockReplace).toHaveBeenCalledWith('/todos?query=mock+filter+text');
        });
    });

    it('should remove the query parameter if the input becomes empty', async () => {
        render(<Searchbar/>);

        const input = screen.getByRole('textbox');
        fireEvent.change(input, {target: {value: ''}});

        await waitFor(() => {
            expect(mockReplace).toHaveBeenCalledWith('/todos');
        });
    });

    it('should navigate to /todos/create if the button was clicked', () => {
        render(<Searchbar/>);

        const link = screen.getByRole('link');
        expect(link).toHaveAttribute('href', '/todos/create');
    });

    it('should match snapshot', () => {
        const {asFragment} = render(<Searchbar/>);
        expect(asFragment()).toMatchSnapshot();
    });
});
