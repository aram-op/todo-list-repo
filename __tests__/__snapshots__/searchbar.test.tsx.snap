// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Searchbar component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <input
      class="mock-poppins-font"
      placeholder="What's up ? ..."
      value="existing"
    />
    <a
      href="/todos/create"
    >
      <button>
        <img
          height="30"
          src="plus.svg"
          width="30"
        />
      </button>
    </a>
  </div>
</DocumentFragment>
`;

exports[`Searchbar component should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <input
      class="mock-poppins-font"
      placeholder="What's up ? ..."
      value="existing"
    />
    <a
      href="/todos/create"
    >
      <button>
        <img
          height="30"
          src="plus.svg"
          width="30"
        />
      </button>
    </a>
  </div>
</DocumentFragment>
`;
